File: erlang/src/staircases.erl
%% ---------------------------------------------------------------------------
%% Copyright (c) 2013 Vladimir Chizhov <master@mirlord.com>
%% This work is licensed under WTFPL v.2 license.
%% See LICENSE and README for details.
%% ---------------------------------------------------------------------------

-module(staircases).
-export([main/1, start/0]).

ncols(Bricks, _, Return) when Bricks =< 0 ->
    Return;

ncols(_, 1, Return) ->
    Return + 1;

ncols(Bricks, Columns, Return) ->
    ncols(Bricks - Columns, Columns,
        ncols(Bricks - Columns, Columns - 1, Return)).

max_cols(Bricks) ->
    round(math:sqrt(Bricks * 2)) + 1.

solve(Bricks) ->
    lists:sum(
        lists:map(
            fun(Columns) ->
                ncols(Bricks, Columns, 0)
            end,
            lists:seq(2, max_cols(Bricks)))).

% Lifecycle & I/O

main([]) ->
    main([io:get_line("")]);

main([Arg | _]) ->
    {Bricks, _} = string:to_integer(Arg),
    io:format("~w~n", [solve(Bricks)]).

start() ->
    main([]).

File: clojure/src/staircases.clj
;; ---------------------------------------------------------------------------
;; Copyright (c) 2013 Vladimir Chizhov <master@mirlord.com>
;; This work is licensed under WTFPL v.2 license.
;; See LICENSE and README for details.
;; ---------------------------------------------------------------------------

(ns staircases)

(defn ^Long ncols
  [^Integer bricks ^Integer columns ^Long return]
  (if (<= bricks 0)
    return
    (if (== columns 1)
      (+ return 1)
      (ncols (- bricks columns)
             columns
             (ncols (- bricks columns)
                    (- columns 1)
                    return)))))

(defn ^Integer max-cols
  [^Integer bricks]
  (+ (Math/round (Math/sqrt (* bricks 2))) 1))

(defn ^Long solve
  [^Integer bricks]
  (apply +
         (map #(ncols bricks % 0)
              (range 2 (max-cols bricks)))))

(defn -main
  [& args]
    (let [bricks (nth args 0)]
      (if bricks
        (println (solve (read-string bricks)))
        (println (solve (read-string (read-line)))))))

